Source/WebCore/Modules/cookie-store/CookieStore.cpp

```diff
@@ -56,25 +65,164 @@ namespace WebCore {
 
 WTF_MAKE_ISO_ALLOCATED_IMPL(CookieStore);
 
-Ref<CookieStore> CookieStore::create(Document* document)
+class CookieStore::MainThreadBridge : public ThreadSafeRefCounted<MainThreadBridge, WTF::DestructionThread::Main> {
+public:
+    static Ref<MainThreadBridge> create(CookieStore& cookieStore, ScriptExecutionContextIdentifier contextIdentifier)
+    {
+        return adoptRef(*new MainThreadBridge(cookieStore, contextIdentifier));
+    }
+
+    void get(CookieStoreGetOptions&&, Function<void(Ref<CookieStore>, ExceptionOr<Vector<Cookie>>&&)>&&);
+    void getAll(CookieStoreGetOptions&&, URL&&, Function<void(Ref<CookieStore>, ExceptionOr<Vector<Cookie>>&&)>&&);
+    void set(CookieInit&& options, Cookie&&, Function<void(Ref<CookieStore>, std::optional<Exception>&&)>&&);
+
+private:
+    MainThreadBridge(CookieStore&, ScriptExecutionContextIdentifier);
+
+    void ensureOnMainThread(Function<void(ScriptExecutionContext&)>&&);
+    void ensureOnContextThread(Function<void()>&&);
+
+    WeakPtr<CookieStore, WeakPtrImplWithEventTargetData> m_cookieStore;
+    ScriptExecutionContextIdentifier m_contextIdentifier;
+};
+
+CookieStore::MainThreadBridge::MainThreadBridge(CookieStore& cookieStore, ScriptExecutionContextIdentifier contextIdentifier)
+    : m_cookieStore(cookieStore)
+    , m_contextIdentifier(contextIdentifier)
 {
-    auto cookieStore = adoptRef(*new CookieStore(document));
-    cookieStore->suspendIfNeeded();
-    return cookieStore;
 }
 
-CookieStore::CookieStore(Document* document)
-    : ActiveDOMObject(document)
+void CookieStore::MainThreadBridge::ensureOnMainThread(Function<void(ScriptExecutionContext&)>&& task)
 {
-    if (!document)
+    ASSERT(m_cookieStore);
+    if (!m_cookieStore)
         return;
 
-    auto* page = document->page();
-    if (!page)
+    RefPtr context = m_cookieStore->scriptExecutionContext();
+    if (!context)
+        return;
+    ASSERT(context->isContextThread());
+
+    if (RefPtr document = dynamicDowncast<Document>(*context)) {
+        task(*context);
         return;
+    }
 
-    m_cookieJar = page->cookieJar();
-    m_host = document->url().host().toString();
+    downcast<WorkerGlobalScope>(*context).thread().workerLoaderProxy().postTaskToLoader(WTFMove(task));
+}
+
+void CookieStore::MainThreadBridge::ensureOnContextThread(Function<void()>&& task)
+{
+    ScriptExecutionContext::ensureOnContextThread(m_contextIdentifier, WTFMove(task));
```

Comment:
```
ScriptExecutionContext::ensureOnContextThread(m_contextIdentifier, [protectedThis = Ref { *this }, task = WTFMove(task)](auto&) {
    if (RefPtr cookieStore = protectedThis->m_cookieStore.get())
        task(*cookieStore);
});
```

Reply:
If we do this, then ensureOnContextThread should look like 
`void ensureOnContextThread(Function<void(CookieStore&)>&&);` instead of 
`void ensureOnContextThread(Function<void(CookieStore*)>&&);` right?

Reply:
Correct, I changed my mind mid-review. I figured there is no point in calling the lambda if the weakPtr is null.